site/app/models/gradeable/Component.php

```diff
@@ -0,0 +1,275 @@
+        $this->setOrder($details['order']);
+        $this->setPage($details['page']);
+    }
+
+    /**
+     * Gets the component's gradeable
+     * @return Gradeable The component's gradeable
+     */
+    public function getGradeable() {
+        return $this->gradeable;
+    }
+
+    /**
+     * Gets the mark object with the provided mark id
+     * @param int $mark_id
+     * @return Mark|null The Mark with the provided id, or null if not found
+     */
+    public function getMark($mark_id) {
+        foreach($this->marks as $mark) {
+            if($mark->getId() === $mark_id) {
+                return $mark;
+            }
+        }
+        return null;
+    }
+
+    /* Overridden setters with validation */
+    /**
+     * Sets the component's gradeable
+     * @param Gradeable $gradeable A non-null gradeable
+     */
+    private function setGradeable(Gradeable $gradeable) {
+        if ($gradeable === null) {
+            throw new \InvalidArgumentException('Gradeable Cannot be null!');
+        }
+        $this->gradeable = $gradeable;
+    }
+
+    const point_properties = [
+        'lower_clamp',
+        'default',
+        'max_value',
+        'upper_clamp'
+    ];
+
+    /**
+     * Parses points from string or float values into all float / null values
+     * @param array $points A partial or complete array of floats or numeric strings indexed by component point property
+     * @return array A complete array of floats (or nulls) indexed by component point property
+     */
+    private function parsePoints(array $points) {
+        $parsedPoints = [];
+        foreach (self::point_properties as $property) {
+            if (is_numeric($points[$property])) {
+                $parsedPoints[$property] = floatval($points[$property]);
+            } else {
+                $parsedPoints[$property] = null;
+            }
+        }
+        return $parsedPoints;
+    }
+
+    /**
+     * Asserts that the point values are valid.  See `setPoints` docs for details
+     *
+     * @param array $points An complete array of floats (or nulls) indexed by component point property
+     */
+    private function assertPoints(array $points) {
+        $errors = array();
+
+        // Give error messages to all null elements
+        foreach (self::point_properties as $property) {
+            if ($points[$property] === null) {
+                $errors[$property] = 'Value must be a number!';
+            }
+        }
+
+        if (Utils::compareNullableGt($points['lower_clamp'], $points['default'])) {
+            $errors['lower_clamp'] = 'Lower clamp can\'t be more than default!';
+        }
+        if (Utils::compareNullableGt($points['default'], $points['max_value'])) {
+            $errors['max_value'] = 'Max value can\'t be less than default!';
+        }
+        if (Utils::compareNullableGt($points['max_value'], $points['upper_clamp'])) {
+            $errors['max_value'] = 'Max value can\'t be more than upper clamp!';
```

Comment:
should this be `upper_clamp` key?

Reply:
It associates the error with `max_value` because `upper_clamp` and `lower_clamp` are supposed to take precedence over `max_value` and `default`